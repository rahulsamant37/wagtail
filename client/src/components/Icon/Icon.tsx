import * as React from 'react';

export interface IconProps extends React.SVGProps<SVGSVGElement> {
  /** Optional svg `path` instead of the `use` based on the icon name. */
  children?: React.ReactNode;
  className?: string;
  name: string;
  title?: string;
}

/**
 * Provide a `title` as an accessible label intended for screen readers.
 */
const Icon: React.FunctionComponent<IconProps> = ({
  children,
  className,
  name,
  title,
  ...props
}) => (
  <>
    <svg
      {...props}
      className={['icon', `icon-${name}`, className || '']
        .filter(Boolean)
        .join(' ')}
      aria-hidden="true"
    >
      {children || <use href={`#icon-${name}`} />}
    </svg>
    {title && <span className="w-sr-only">{title}</span>}
  </>
);

export default Icon;
