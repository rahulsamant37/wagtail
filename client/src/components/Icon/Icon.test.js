import React from 'react';
import { shallow, mount } from 'enzyme';
import Icon from './Icon';

describe('Icon', () => {
  it('exists', () => {
    expect(Icon).toBeDefined();
  });

  it('should support icons with a name', () => {
    const wrapper = mount(<Icon name="test" />);

    expect(wrapper.find('.icon.icon-test')).toHaveLength(1);
    expect(wrapper.find('use[href="#icon-test"]')).toHaveLength(1);

    expect(wrapper).toMatchSnapshot();
  });

  it('should support children in place of the icon use#name', () => {
    const wrapper = shallow(
      <icon name="example">
        <rect
          x="10"
          y="10"
          width="30"
          height="30"
          stroke="black"
          fill="transparent"
          strokeWidth="5"
        />
      </icon>,
    );

    expect(wrapper.find('use')).toHaveLength(0);
    expect(wrapper.find('rect')).toHaveLength(1);
  });

  it('should support a className prop', () => {
    const wrapper = mount(<Icon name="test" className="u-test" />);

    expect(wrapper.find('.icon.u-test')).toHaveLength(1);
  });

  it('should support other svg attributes', () => {
    const wrapper = mount(<Icon name="test" viewBox="0 0 1024 1024" />);

    expect(wrapper.find('svg').prop('viewBox')).toBe('0 0 1024 1024');
  });

  it('should support a title that is output as a sibling of the title', () => {
    const wrapper = mount(<Icon name="test" title="Test title" />);

    const title = wrapper.find('svg.icon ~ span');
    expect(title).toHaveLength(1);

    expect(title.text()).toBe('Test title');
    expect(title.hasClass('w-sr-only')).toBe(true);
  });
});
