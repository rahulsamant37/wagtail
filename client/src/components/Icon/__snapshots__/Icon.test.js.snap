// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should support icons with a name 1`] = `
<Icon
  name="test"
>
  <svg
    aria-hidden="true"
    className="icon icon-test"
  >
    <use
      href="#icon-test"
    />
  </svg>
</Icon>
`;
